File: app/src/main/java/com/thibaudperso/sonycamera/timelapse/model/ApiRequestsList.java
package com.thibaudperso.sonycamera.timelapse.model;

import com.thibaudperso.sonycamera.sdk.model.PictureResponse;

import java.util.LinkedHashMap;

public class ApiRequestsList extends LinkedHashMap<Long, PictureResponse> {

    /*
     * Pre-process attributes
     */
    private int mRequestsSent;
    private int mResponsesReceived;

    private int mResponsesWsUnreachable;
    private int mResponsesLongShot;
    private int mResponsesUnknown;

    private long mLastRequestSent;

    private String mLastPictureUrl;

    ApiRequestsList() {
        initPreProcessAttributes();
    }

    private void initPreProcessAttributes() {

        mRequestsSent = 0;
        mResponsesReceived = 0;
        mResponsesWsUnreachable = 0;
        mResponsesLongShot = 0;
        mResponsesUnknown = 0;
        mLastRequestSent = 0;
        mLastPictureUrl = null;
    }

    @Override
    public void clear() {
        super.clear();
        initPreProcessAttributes();
    }

    @Override
    public PictureResponse put(Long key, PictureResponse value) {

        if (value == null) {
            mRequestsSent++;
            mLastRequestSent = key;
        } else {
            mResponsesReceived++;

            switch (value.status) {
                case OK:
                    mLastPictureUrl = value.url;
                    break;

                case WS_UNREACHABLE:
                    mResponsesWsUnreachable++;
                    break;

                case LONG_SHOOTING:
                    mResponsesLongShot++;
                    break;

                default:
                    mResponsesUnknown++;
            }

        }
        return super.put(key, value);
    }


    public int getRequestsSent() {
        return mRequestsSent;
    }

    public int getResponsesReceived() {
        return mResponsesReceived;
    }

    public int getResponsesWsUnreachable() {
        return mResponsesWsUnreachable;
    }

    public int getResponsesLongShot() {
        return mResponsesLongShot;
    }

    public int getResponsesUnknown() {
        return mResponsesUnknown;
    }

    public int getNumberOfSkippedFrames() {
        return mResponsesLongShot + mResponsesWsUnreachable + mResponsesUnknown;
    }

    public long getLastRequestSent() {
        return mLastRequestSent;
    }

    public String getLastPictureUrl() {
        return mLastPictureUrl;
    }

    public boolean isTakingPicture() {
        return mRequestsSent != mResponsesReceived;
    }

    @Override
    public String toString() {
        return "ApiRequestsList{" +
                "mRequestsSent=" + mRequestsSent +
                ", mResponsesReceived=" + mResponsesReceived +
                ", mResponsesWsUnreachable=" + mResponsesWsUnreachable +
                ", mResponsesLongShot=" + mResponsesLongShot +
                ", mResponsesUnknown=" + mResponsesUnknown +
                '}';
    }

}

File: app/src/main/java/com/thibaudperso/sonycamera/timelapse/model/ApiRequestsList.kt
package com.thibaudperso.sonycamera.timelapse.model

import com.thibaudperso.sonycamera.sdk.ResponseCode
import com.thibaudperso.sonycamera.sdk.model.PictureResponse
import java.util.*

class ApiRequestsList : LinkedHashMap<Long, PictureResponse?>() {

    var requestsSent: Int = 0
        private set
    var responsesReceived: Int = 0
        private set

    var responsesWsUnreachable: Int = 0
        private set
    var responsesLongShot: Int = 0
        private set
    var responsesUnknown: Int = 0
        private set

    var lastRequestSent: Long = 0
        private set

    var lastPictureUrl: String? = null
        private set

    val numberOfSkippedFrames: Int
        get() = responsesLongShot + responsesWsUnreachable + responsesUnknown

    val isTakingPicture: Boolean
        get() = requestsSent != responsesReceived

    init {
        initPreProcessAttributes()
    }

    override fun clear() {
        super.clear()
        initPreProcessAttributes()
    }

    override fun put(key: Long, value: PictureResponse?): PictureResponse? {
        if (value == null) {
            requestsSent++
            lastRequestSent = key
        } else {
            responsesReceived++
            when (value.status) {
                ResponseCode.OK -> lastPictureUrl = value.url
                ResponseCode.WS_UNREACHABLE -> responsesWsUnreachable++
                ResponseCode.LONG_SHOOTING -> responsesLongShot++
                else -> responsesUnknown++
            }
        }
        return super.put(key, value)
    }

    override fun toString(): String = "ApiRequestsList{requestsSent=$requestsSent, responsesReceived=$responsesReceived, responsesWsUnreachable=$responsesWsUnreachable, responsesLongShot=$responsesLongShot, responsesUnknown=$responsesUnknown}"

    private fun initPreProcessAttributes() {
        requestsSent = 0
        responsesReceived = 0
        responsesWsUnreachable = 0
        responsesLongShot = 0
        responsesUnknown = 0
        lastRequestSent = 0
        lastPictureUrl = null
    }

}
